### ISSUE TITLE
Scheduler's snapshot.nodeInfoList has data chaos

### ISSUE THREAD

**Issue by coderwangke** [kubernetes/kubernetes#97120]
**What happened**:
k8s cluster has for node (3 for master node, and one for worker node), basic info:
master-01: label[failure-domain.beta.kubernetes.io/region=sh,failure-domain.beta.kubernetes.io/zone=200002]
master-02: label[failure-domain.beta.kubernetes.io/region=sh,failure-domain.beta.kubernetes.io/zone=200002]
master-03: label[failure-domain.beta.kubernetes.io/region=sh,failure-domain.beta.kubernetes.io/zone=200002]
worker-node-01: label[failure-domain.beta.kubernetes.io/region=sh,failure-domain.beta.kubernetes.io/zone=200004]

when create a daemonset, one pod stuck in pending state：
NAME          READY   STATUS    RESTARTS   AGE
debug-4m8lc   1/1     Running   1          89m
debug-dn47c   0/1     Pending   0          89m
debug-lkmfs   1/1     Running   1          89m
debug-qwdbc   1/1     Running   1          89m
**What you expected to happen**:
no pod stuck in pending state
**How to reproduce it (as minimally and precisely as possible)**:
create a cluster with 3 master and 1 worker node (worker node and master are in different zones）, and create a daemonset.
```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: debug
  labels:
    k8s-app: debug
spec:
  selector:
    matchLabels:
      name: debug
  template:
    metadata:
      labels:
        name: debug
    spec:
      tolerations:
      - operator: Exists
      containers:
      - name: debug
        image: busybox
        command: ["sleep", "1h"]
      terminationGracePeriodSeconds: 30
```

**Anything else we need to know?**:
we have known the reason.
#### first round:
After k8s cluster creates(with 3 master), when scheduling pod (maybe from daemonset, eg: cni plugin), there is a Sync，from cache.Nodetree to snapshot.nodeInfoList, and then, We get the following result in cache:
 nodeTree.zoneIndex = 1,
 nodeTree.nodeArray[sh:200002].lastIndex = 3,
#### second round:
when worker node added to k8s cluster, and then create a daemonset, there is a Sync too, from cache.Nodetree to snapshot.nodeInfoList, The process is as follows:
1. zoneIndex =1, nodeArray[sh:200004].lastIndex = 0, we get worker-node-01.
2. zoneIndex = 2 > len(zones); zoneIndex = 0, nodeArray[sh:200002].lastIndex = 3, return.
3. zoneIndex = 1, nodeArray[sh:200004].lastIndex = 1, return.
4. zoneIndex = 0, nodeArray[sh:200002].lastIndex = 0, we get master-01.
5. zoneIndex = 1, nodeArray[sh:200004].lastIndex = 0, we get worker-node-01.
6. zoneIndex = 2 > len(zones); zoneIndex = 0, nodeArray[sh:200002].lastIndex = 1, we get master-02.

After Sync, we get a snapshot.nodeInfoList with following info:
[
worker-node-01,
master-01,
worker-node-01,
master-02,
].
And where is master-03?

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.4", GitCommit:"f8797eb035aca9c928af02ea9a74d232cfd41bc6", GitTreeState:"clean", BuildDate:"2020-10-12T03:38:05Z", GoVersion:"go1.14.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.4", GitCommit:"f8797eb035aca9c928af02ea9a74d232cfd41bc6", GitTreeState:"dirty", BuildDate:"2020-12-08T10:56:38Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}

**Comment by Huang-Wei** [comment]
@coderwangke this issue should have been fixed in releases >= **1.19** (by #93355), and backported to **1.18.8** and onwards (by #93387). Please let me know if you still see the issue on the latest 1.18 patches.

**Comment by coderwangke** [comment]
@Huang-Wei Thanks for your comment, i will merge #93387 to v1.18 and test again.
